// ng-busy/esm5/lib/ng-busy.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { BusyConfig, DefaultBusyComponent } from './model/busy-config';
import { CommonModule } from '@angular/common';
import { BusyTrackerService } from './service/busy-tracker.service';
import { BusyConfigHolderService } from './service/busy-config-holder.service';
import { NgBusyDirective } from './ng-busy.directive';
import { NgBusyComponent } from './component/ng-busy/ng-busy.component';
var NgBusyModule = /** @class */ (function () {
    function NgBusyModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    NgBusyModule.forRoot = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: NgBusyModule,
            providers: [
                { provide: BusyConfig, useValue: config }
            ]
        };
    };
    NgBusyModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    declarations: [DefaultBusyComponent, NgBusyDirective, NgBusyComponent],
                    providers: [BusyConfigHolderService, BusyTrackerService],
                    exports: [NgBusyDirective],
                    entryComponents: [DefaultBusyComponent, NgBusyComponent]
                },] },
    ];
    return NgBusyModule;
}());
export { NgBusyModule };
