// ng-busy/esm5/lib/model/busy-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Inject } from '@angular/core';
import { InstanceConfigHolderService } from '../service/instance-config-holder.service';
var BusyConfig = /** @class */ (function () {
    function BusyConfig(config) {
        var e_1, _a;
        if (config === void 0) { config = {}; }
        try {
            for (var _b = tslib_1.__values(Object.keys(BUSY_CONFIG_DEFAULTS)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var option = _c.value;
                this[option] = config[option] !== undefined ? config[option] : BUSY_CONFIG_DEFAULTS[option];
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    return BusyConfig;
}());
export { BusyConfig };
if (false) {
    /** @type {?} */
    BusyConfig.prototype.template;
    /** @type {?} */
    BusyConfig.prototype.templateNgStyle;
    /** @type {?} */
    BusyConfig.prototype.delay;
    /** @type {?} */
    BusyConfig.prototype.minDuration;
    /** @type {?} */
    BusyConfig.prototype.backdrop;
    /** @type {?} */
    BusyConfig.prototype.message;
    /** @type {?} */
    BusyConfig.prototype.wrapperClass;
    /** @type {?} */
    BusyConfig.prototype.disableAnimation;
}
var DefaultBusyComponent = /** @class */ (function () {
    function DefaultBusyComponent(instanceConfigHolder) {
        this.instanceConfigHolder = instanceConfigHolder;
    }
    Object.defineProperty(DefaultBusyComponent.prototype, "message", {
        get: /**
         * @return {?}
         */
        function () {
            return this.instanceConfigHolder.config.message;
        },
        enumerable: true,
        configurable: true
    });
    DefaultBusyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'default-busy',
                    template: "\n      <div class=\"ng-busy-default-wrapper\">\n          <div class=\"ng-busy-default-sign\">\n              <div class=\"ng-busy-default-spinner\">\n                  <div class=\"bar1\"></div>\n                  <div class=\"bar2\"></div>\n                  <div class=\"bar3\"></div>\n                  <div class=\"bar4\"></div>\n                  <div class=\"bar5\"></div>\n                  <div class=\"bar6\"></div>\n                  <div class=\"bar7\"></div>\n                  <div class=\"bar8\"></div>\n                  <div class=\"bar9\"></div>\n                  <div class=\"bar10\"></div>\n                  <div class=\"bar11\"></div>\n                  <div class=\"bar12\"></div>\n              </div>\n              <div class=\"ng-busy-default-text\">{{message}}</div>\n          </div>\n      </div>\n  ",
                },] },
    ];
    /** @nocollapse */
    DefaultBusyComponent.ctorParameters = function () { return [
        { type: InstanceConfigHolderService, decorators: [{ type: Inject, args: ['instanceConfigHolder',] }] }
    ]; };
    return DefaultBusyComponent;
}());
export { DefaultBusyComponent };
if (false) {
    /** @type {?} */
    DefaultBusyComponent.prototype.instanceConfigHolder;
}
/**
 * @record
 */
export function IBusyConfig() { }
/** @type {?|undefined} */
IBusyConfig.prototype.template;
/** @type {?|undefined} */
IBusyConfig.prototype.templateNgStyle;
/** @type {?|undefined} */
IBusyConfig.prototype.delay;
/** @type {?|undefined} */
IBusyConfig.prototype.minDuration;
/** @type {?|undefined} */
IBusyConfig.prototype.backdrop;
/** @type {?|undefined} */
IBusyConfig.prototype.message;
/** @type {?|undefined} */
IBusyConfig.prototype.wrapperClass;
/** @type {?|undefined} */
IBusyConfig.prototype.busy;
/** @type {?|undefined} */
IBusyConfig.prototype.disableAnimation;
/** @type {?} */
export var BUSY_CONFIG_DEFAULTS = {
    template: DefaultBusyComponent,
    templateNgStyle: {},
    delay: 0,
    minDuration: 0,
    backdrop: true,
    message: 'Please wait...',
    wrapperClass: 'ng-busy',
    disableAnimation: false
};
