// ng-busy/esm2015/lib/component/ng-busy/ng-busy.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, EventEmitter, Inject } from '@angular/core';
import { animate, style, transition, trigger } from '@angular/animations';
import { InstanceConfigHolderService } from '../../service/instance-config-holder.service';
/** @type {?} */
const inactiveStyle = style({
    opacity: 0,
    transform: 'translateY(-40px)'
});
/** @type {?} */
const timing = '.3s ease';
export class NgBusyComponent {
    /**
     * @param {?} instanceConfigHolder
     * @param {?} busyEmitter
     * @param {?} cdr
     */
    constructor(instanceConfigHolder, busyEmitter, cdr) {
        this.instanceConfigHolder = instanceConfigHolder;
        this.busyEmitter = busyEmitter;
        this.cdr = cdr;
        this.disableAnimation = false;
        this.showBackdrop = true;
        this.isActive = false;
        this.busyMonitor = this.busyEmitter.subscribe((isActive) => {
            /** @type {?} */
            const config = this.instanceConfigHolder.config;
            this.isActive = isActive;
            this.wrapperClass = config.wrapperClass;
            this.showBackdrop = config.backdrop;
            this.disableAnimation = config.disableAnimation;
            if (this.cdr) {
                this.cdr.markForCheck();
            }
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.busyMonitor) {
            this.busyMonitor.unsubscribe();
        }
    }
}
NgBusyComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-ng-busy',
                template: `<div [class]="wrapperClass" @flyInOut [@.disabled]="disableAnimation" *ngIf="isActive">
  <ng-content></ng-content>
</div>
<div class="ng-busy-backdrop" @flyInOut [@.disabled]="disableAnimation" *ngIf="showBackdrop && isActive">
</div>
`,
                styles: [``],
                animations: [
                    trigger('flyInOut', [
                        transition('void => *', [
                            inactiveStyle,
                            animate(timing)
                        ]),
                        transition('* => void', [
                            animate(timing, inactiveStyle)
                        ])
                    ])
                ]
            },] },
];
/** @nocollapse */
NgBusyComponent.ctorParameters = () => [
    { type: InstanceConfigHolderService, decorators: [{ type: Inject, args: ['instanceConfigHolder',] }] },
    { type: EventEmitter, decorators: [{ type: Inject, args: ['busyEmitter',] }] },
    { type: ChangeDetectorRef }
];
if (false) {
    /** @type {?} */
    NgBusyComponent.prototype.wrapperClass;
    /** @type {?} */
    NgBusyComponent.prototype.disableAnimation;
    /** @type {?} */
    NgBusyComponent.prototype.showBackdrop;
    /** @type {?} */
    NgBusyComponent.prototype.busyMonitor;
    /** @type {?} */
    NgBusyComponent.prototype.isActive;
    /** @type {?} */
    NgBusyComponent.prototype.instanceConfigHolder;
    /** @type {?} */
    NgBusyComponent.prototype.busyEmitter;
    /** @type {?} */
    NgBusyComponent.prototype.cdr;
}
