// ng-busy/esm5/lib/service/busy-tracker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { from, timer } from 'rxjs';
import { isPromise } from '../util/isPromise';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function TrackerOptions() { }
/** @type {?} */
TrackerOptions.prototype.minDuration;
/** @type {?} */
TrackerOptions.prototype.delay;
/** @type {?} */
TrackerOptions.prototype.busyList;
var BusyTrackerService = /** @class */ (function () {
    function BusyTrackerService() {
        this.isDelayProcessing = false;
        this.isDurationProcessing = false;
        this.isBusiesProcessing = false;
        this.busyQueue = [];
        this.__isActive = false;
        this.onStartBusy = new EventEmitter();
        this.onStopBusy = new EventEmitter();
    }
    Object.defineProperty(BusyTrackerService.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this.__isActive;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (this.__isActive === false && val === true && this.onStartBusy) {
                this.onStartBusy.emit();
            }
            if (this.__isActive === true && val === false && this.onStopBusy) {
                this.isBusiesProcessing = false;
                this.onStopBusy.emit();
            }
            this.__isActive = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BusyTrackerService.prototype, "busyList", {
        get: /**
         * @return {?}
         */
        function () {
            return this.busyQueue;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} options
     * @return {?}
     */
    BusyTrackerService.prototype.load = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        this.loadBusyQueue(options.busyList);
        this.startLoading(options);
    };
    /**
     * @return {?}
     */
    BusyTrackerService.prototype.updateActiveStatus = /**
     * @return {?}
     */
    function () {
        this.isActive = this.isBusiesProcessing &&
            !this.isDelayProcessing &&
            (this.isDurationProcessing || this.busyQueue.length > 0);
    };
    /**
     * @param {?} options
     * @return {?}
     */
    BusyTrackerService.prototype.startLoading = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        var _this = this;
        if (!this.isBusiesProcessing && this.busyList.length > 0) {
            this.isBusiesProcessing = true;
            this.isDelayProcessing = true;
            this.updateActiveStatus();
            timer(options.delay).subscribe(function () {
                _this.isDelayProcessing = false;
                _this.isDurationProcessing = true;
                _this.updateActiveStatus();
                timer(options.minDuration).subscribe(function () {
                    _this.isDurationProcessing = false;
                    _this.updateActiveStatus();
                });
            });
        }
    };
    /**
     * @param {?} busies
     * @return {?}
     */
    BusyTrackerService.prototype.loadBusyQueue = /**
     * @param {?} busies
     * @return {?}
     */
    function (busies) {
        var _this = this;
        busies.filter(function (busy) {
            return busy && !busy.hasOwnProperty('__loaded_mark_by_ng_busy');
        }).forEach(function (busy) {
            Object.defineProperty(busy, '__loaded_mark_by_ng_busy', {
                value: true, configurable: false, enumerable: false, writable: false
            });
            /** @type {?} */
            var cur_busy;
            if (isPromise(busy)) {
                cur_busy = from(busy).subscribe();
            }
            else {
                cur_busy = busy;
            }
            _this.appendToQueue(cur_busy);
        });
    };
    /**
     * @param {?} busy
     * @return {?}
     */
    BusyTrackerService.prototype.appendToQueue = /**
     * @param {?} busy
     * @return {?}
     */
    function (busy) {
        var _this = this;
        this.busyQueue.push(busy);
        busy.add(function () {
            _this.busyQueue = _this.busyQueue.filter(function (cur) { return !cur.closed; });
            _this.updateActiveStatus();
        });
    };
    /**
     * @return {?}
     */
    BusyTrackerService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    BusyTrackerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    BusyTrackerService.ctorParameters = function () { return []; };
    /** @nocollapse */ BusyTrackerService.ngInjectableDef = i0.defineInjectable({ factory: function BusyTrackerService_Factory() { return new BusyTrackerService(); }, token: BusyTrackerService, providedIn: "root" });
    return BusyTrackerService;
}());
export { BusyTrackerService };
if (false) {
    /** @type {?} */
    BusyTrackerService.prototype.isDelayProcessing;
    /** @type {?} */
    BusyTrackerService.prototype.isDurationProcessing;
    /** @type {?} */
    BusyTrackerService.prototype.isBusiesProcessing;
    /** @type {?} */
    BusyTrackerService.prototype.busyQueue;
    /** @type {?} */
    BusyTrackerService.prototype.__isActive;
    /** @type {?} */
    BusyTrackerService.prototype.onStartBusy;
    /** @type {?} */
    BusyTrackerService.prototype.onStopBusy;
}
