// ng-busy/esm2015/lib/service/busy-tracker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { from, timer } from 'rxjs';
import { isPromise } from '../util/isPromise';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function TrackerOptions() { }
/** @type {?} */
TrackerOptions.prototype.minDuration;
/** @type {?} */
TrackerOptions.prototype.delay;
/** @type {?} */
TrackerOptions.prototype.busyList;
export class BusyTrackerService {
    constructor() {
        this.isDelayProcessing = false;
        this.isDurationProcessing = false;
        this.isBusiesProcessing = false;
        this.busyQueue = [];
        this.__isActive = false;
        this.onStartBusy = new EventEmitter();
        this.onStopBusy = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get isActive() {
        return this.__isActive;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set isActive(val) {
        if (this.__isActive === false && val === true && this.onStartBusy) {
            this.onStartBusy.emit();
        }
        if (this.__isActive === true && val === false && this.onStopBusy) {
            this.isBusiesProcessing = false;
            this.onStopBusy.emit();
        }
        this.__isActive = val;
    }
    /**
     * @return {?}
     */
    get busyList() {
        return this.busyQueue;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    load(options) {
        this.loadBusyQueue(options.busyList);
        this.startLoading(options);
    }
    /**
     * @return {?}
     */
    updateActiveStatus() {
        this.isActive = this.isBusiesProcessing &&
            !this.isDelayProcessing &&
            (this.isDurationProcessing || this.busyQueue.length > 0);
    }
    /**
     * @param {?} options
     * @return {?}
     */
    startLoading(options) {
        if (!this.isBusiesProcessing && this.busyList.length > 0) {
            this.isBusiesProcessing = true;
            this.isDelayProcessing = true;
            this.updateActiveStatus();
            timer(options.delay).subscribe(() => {
                this.isDelayProcessing = false;
                this.isDurationProcessing = true;
                this.updateActiveStatus();
                timer(options.minDuration).subscribe(() => {
                    this.isDurationProcessing = false;
                    this.updateActiveStatus();
                });
            });
        }
    }
    /**
     * @param {?} busies
     * @return {?}
     */
    loadBusyQueue(busies) {
        busies.filter((busy) => {
            return busy && !busy.hasOwnProperty('__loaded_mark_by_ng_busy');
        }).forEach((busy) => {
            Object.defineProperty(busy, '__loaded_mark_by_ng_busy', {
                value: true, configurable: false, enumerable: false, writable: false
            });
            /** @type {?} */
            let cur_busy;
            if (isPromise(busy)) {
                cur_busy = from(busy).subscribe();
            }
            else {
                cur_busy = busy;
            }
            this.appendToQueue(cur_busy);
        });
    }
    /**
     * @param {?} busy
     * @return {?}
     */
    appendToQueue(busy) {
        this.busyQueue.push(busy);
        busy.add(() => {
            this.busyQueue = this.busyQueue.filter((cur) => !cur.closed);
            this.updateActiveStatus();
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
BusyTrackerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
BusyTrackerService.ctorParameters = () => [];
/** @nocollapse */ BusyTrackerService.ngInjectableDef = i0.defineInjectable({ factory: function BusyTrackerService_Factory() { return new BusyTrackerService(); }, token: BusyTrackerService, providedIn: "root" });
if (false) {
    /** @type {?} */
    BusyTrackerService.prototype.isDelayProcessing;
    /** @type {?} */
    BusyTrackerService.prototype.isDurationProcessing;
    /** @type {?} */
    BusyTrackerService.prototype.isBusiesProcessing;
    /** @type {?} */
    BusyTrackerService.prototype.busyQueue;
    /** @type {?} */
    BusyTrackerService.prototype.__isActive;
    /** @type {?} */
    BusyTrackerService.prototype.onStartBusy;
    /** @type {?} */
    BusyTrackerService.prototype.onStopBusy;
}
