// ng-busy/esm2015/lib/model/busy-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Inject } from '@angular/core';
import { InstanceConfigHolderService } from '../service/instance-config-holder.service';
export class BusyConfig {
    /**
     * @param {?=} config
     */
    constructor(config = {}) {
        for (const option of Object.keys(BUSY_CONFIG_DEFAULTS)) {
            this[option] = config[option] !== undefined ? config[option] : BUSY_CONFIG_DEFAULTS[option];
        }
    }
}
if (false) {
    /** @type {?} */
    BusyConfig.prototype.template;
    /** @type {?} */
    BusyConfig.prototype.templateNgStyle;
    /** @type {?} */
    BusyConfig.prototype.delay;
    /** @type {?} */
    BusyConfig.prototype.minDuration;
    /** @type {?} */
    BusyConfig.prototype.backdrop;
    /** @type {?} */
    BusyConfig.prototype.message;
    /** @type {?} */
    BusyConfig.prototype.wrapperClass;
    /** @type {?} */
    BusyConfig.prototype.disableAnimation;
}
export class DefaultBusyComponent {
    /**
     * @param {?} instanceConfigHolder
     */
    constructor(instanceConfigHolder) {
        this.instanceConfigHolder = instanceConfigHolder;
    }
    /**
     * @return {?}
     */
    get message() {
        return this.instanceConfigHolder.config.message;
    }
}
DefaultBusyComponent.decorators = [
    { type: Component, args: [{
                selector: 'default-busy',
                template: `
      <div class="ng-busy-default-wrapper">
          <div class="ng-busy-default-sign">
              <div class="ng-busy-default-spinner">
                  <div class="bar1"></div>
                  <div class="bar2"></div>
                  <div class="bar3"></div>
                  <div class="bar4"></div>
                  <div class="bar5"></div>
                  <div class="bar6"></div>
                  <div class="bar7"></div>
                  <div class="bar8"></div>
                  <div class="bar9"></div>
                  <div class="bar10"></div>
                  <div class="bar11"></div>
                  <div class="bar12"></div>
              </div>
              <div class="ng-busy-default-text">{{message}}</div>
          </div>
      </div>
  `,
            },] },
];
/** @nocollapse */
DefaultBusyComponent.ctorParameters = () => [
    { type: InstanceConfigHolderService, decorators: [{ type: Inject, args: ['instanceConfigHolder',] }] }
];
if (false) {
    /** @type {?} */
    DefaultBusyComponent.prototype.instanceConfigHolder;
}
/**
 * @record
 */
export function IBusyConfig() { }
/** @type {?|undefined} */
IBusyConfig.prototype.template;
/** @type {?|undefined} */
IBusyConfig.prototype.templateNgStyle;
/** @type {?|undefined} */
IBusyConfig.prototype.delay;
/** @type {?|undefined} */
IBusyConfig.prototype.minDuration;
/** @type {?|undefined} */
IBusyConfig.prototype.backdrop;
/** @type {?|undefined} */
IBusyConfig.prototype.message;
/** @type {?|undefined} */
IBusyConfig.prototype.wrapperClass;
/** @type {?|undefined} */
IBusyConfig.prototype.busy;
/** @type {?|undefined} */
IBusyConfig.prototype.disableAnimation;
/** @type {?} */
export const BUSY_CONFIG_DEFAULTS = {
    template: DefaultBusyComponent,
    templateNgStyle: {},
    delay: 0,
    minDuration: 0,
    backdrop: true,
    message: 'Please wait...',
    wrapperClass: 'ng-busy',
    disableAnimation: false
};
